extensions/parquet/table/src/main/java/io/deephaven/parquet/table/util/S3SeekableByteChannel.java

```diff
@@ -0,0 +1,273 @@
+     * The size of the object in bytes
+     */
+    private final long size;
+
+    /**
+     * The maximum time and units to wait while fetching an object
+     **/
+    private final Long timeout;
+    private final TimeUnit timeUnit;
+
+
+    S3SeekableByteChannel(@NotNull final SeekableChannelsProvider.ChannelContext context, @NotNull final String bucket,
+                          @NotNull final String key, @NotNull final S3AsyncClient s3Client, final long size,
+                          final int fragmentSize, final int readAheadCount) {
+        this.position = 0;
+        this.bucket = bucket;
+        this.key = key;
+        this.s3Client = s3Client;
+
+        this.fragmentSize = fragmentSize;
+        this.readAheadCount = readAheadCount;
+        this.timeout = CONNECTION_TIMEOUT_MINUTES;
+        this.timeUnit = CONNECTION_TIMEOUT_UNIT;
+        this.size = size;
+        this.numFragmentsInObject = (int) Math.ceil((double) size / fragmentSize);
+        this.context = (S3ChannelContext) context;
+        this.finalFragmentIdx = numFragmentsInObject - 1;
+        this.finalFragment = null;
+    }
+
+    @Override
+    public void setContext(@Nullable final SeekableChannelsProvider.ChannelContext context) {
+        // null context equivalent to clearing the context
+        if (context != null && !(context instanceof S3ChannelContext)) {
+            throw new IllegalArgumentException("context must be null or an instance of ChannelContext, provided context " +
+                    " of class " + context.getClass().getName());
+        }
+        this.context = (S3ChannelContext) context;
+    }
+
+    @Override
+    public int read(@NotNull final ByteBuffer destination) throws ClosedChannelException {
+        Assert.neqNull(context, "context");
+        if (!destination.hasRemaining()) {
+            return 0;
+        }
+        final long localPosition = position;
+        checkClosed(localPosition);
+        if (localPosition >= size) {
+            // We are finished reading
+            return -1;
+        }
+
+        final int currFragmentIndex = fragmentIndexForByteNumber(localPosition);
+        final ByteBuffer currentFragment;
+        if (currFragmentIndex == finalFragmentIdx && finalFragment != null) {
+            currentFragment = finalFragment.asReadOnlyBuffer();
+        } else {
+            // Send async read request the current fragment, if it's not already in the cache
+            final S3ChannelContext.FragmentContext fragmentContext = context.getFragmentContext(currFragmentIndex);
+            final CompletableFuture<ByteBuffer> fetchCurrFragment;
+            if (fragmentContext != null && fragmentContext.fragmentIndex == currFragmentIndex) {
+                fetchCurrFragment = fragmentContext.future;
+            } else {
+                fetchCurrFragment = computeFragmentFuture(currFragmentIndex);
+                context.setFragmentContext(currFragmentIndex, fetchCurrFragment);
+            }
```

Comment:
I wonder if this logic could be removed, and we just do it based on current + read-ahead priming below?

Reply:
This logic is not corrects since I am keeping a local copy in the channel instead of sharing it across instances.
So I have removed this logic now.